### ISSUE TITLE
Expose disk_interface to python

### ISSUE THREAD

**Issue by KOLANICH** [arvidn/libtorrent#5296]
It would be nice to have `disk_interface` exposed to python. The goal is measuring internet speed without writing a byte to storage.

**Comment by arvidn** [comment]
there are challenges with calling python functions from the libtorrent thread. Either you get race conditions not locking the GIL, or you get deadlocks because the python thread holds the GIL and is waiting for the network thread, when the network thread tries to lock it.

**Comment by arvidn** [comment]
I don't see how this can be done without first solving the python threading issue. Do you have any ideas of *how* this could be exposed to python?

**Comment by KOLANICH** [comment]
Could you, please, put the interfaces (and maybe threads somehow) involved to https://github.com/arvidn/libtorrent/blob/RC_2_0/docs/img/hacking.diagram ?

**Comment by arvidn** [comment]
by "the interfaces", do you mean `disk_interface`?

**Comment by KOLANICH** [comment]
`disk_interface` and all the other interfaces and their interactions I must understand in order to understand the issue. Because currently I don't understand why GIL would cause deadlocks and if it is really needed for python code be executed in both threads simultaniously. If it is not needed, I currently see no problem at all, if it is needed, but the code must run simultaniously in some synchronized way, I wonder if cooperative multitasking can be used anyhow. If even a cooperative multitasking cannot be used, I wonder if the network thread can be moved (if needed) into another process with interactions between them through a shared memory mapping.

**Comment by arvidn** [comment]
> currently I don't understand why GIL would cause deadlocks and if it is really needed for python code be executed in both threads simultaniously.

the GIL must always be held when executing python code, the interpreter has global state that may not be accessed simultaneously by multiple threads. This is why python doesn't really support multi-threading, normally you use multiple processes in python instead, for this reason. The thing that *can* be parallellized with threads in python are blocking system calls, that may have released the GIL while waiting.

> I wonder if the network thread can be moved (if needed) into another process with interactions between them through a shared memory mapping.

It cannot (it would be quite a lot of work to do so at least). That process would still need a python interpreter associated with it, and the python callbacks could not share any state with the original python interpreter.

Normally data structures don't trivially support living in shared memory, you need specialized data structures for that, with fancy pointers.

**Comment by KOLANICH** [comment]
Yeah, I know what GIL is and that it is locked while python code is running. I just don't understand the architecture of this lib enough.

I mean, to have a deadlock we need to have disk_interface thread waiting for a sync primitive in the network thread, and one in the network thread was waiting for GIL in `disk_interface` thread so cannot do its work and satisfy the conditions for disk thread. But how are these threads interacting? Why do we have to have the GIL locked in disk_interface thread while it is sleeping? (Is my mental model of how this lib works correct (I haven't read the code yet) ?)

**Comment by arvidn** [comment]
My initial reaction was based on experience trying to expose extension plugins to python, which run in the main libtorrent network thread. That thread definitely causes dead-locks, because there are blocking API calls that post work to it and wait for it to complete.

The only way I can think of the disk interface causing a deadlock is if the `aio_threads` setting is 0, in which case all disk I/O happens in the network thread as well.

**Comment by KOLANICH** [comment]
>The only way I can think of the disk interface causing a deadlock is if the aio_threads setting is 0, in which case all disk I/O happens in the network thread as well.

Why are any deadlocks possible when the critical section are in the same thread?

**Comment by arvidn** [comment]
> Why are any deadlocks possible when the critical sections are in the same thread?

The python (main) thread and libtorrent's network thread are not the same thread. If there are blocking python calls (waiting for the network thread) and the network thread can also be waiting on the GIL, there may be a deadlock.

**Comment by AllSeeingEyeTolledEweSew** [comment]
Beyond what's discussed above, you might also consider that python is slow, and implementing a low-level interface there probably isn't the best way to do performance testing.

I often test internet speeds with bittorrent on a tmpfs. Would that work for you?

**Comment by KOLANICH** [comment]
>on tmpfs

means wasting memory

**Comment by AllSeeingEyeTolledEweSew** [comment]
> means wasting memory

You have to store the data somewhere, so pieces can be hashed when all blocks are downloaded. It must either be in memory or disk. On average the working set *should* be confined to a few pieces and you could discard the others once hashed, but you'd need to deal with the case when that's not true (i.e. you've downloaded all blocks but one, from every piece). This idea starts to look less like the bittorrent protocol the more I think about it, so I'm not sure libtorrent is the right choice for you.

I'm definitely sure python isn't the right choice for you. If you're worried that storing a test torrent takes up too much memory, you're likely working at network speeds that would be hindered by python being slow.

Data transfer callbacks should just not be done in python, and I say this as someone who loves python pathologically.
